File: LuoGu/P1518.cpp
//#include <iostream>
//
//using namespace std;
//
//// 这题的精髓是通过保存状态来解决" 如果 John 和牛永远不会相遇"这一条件--》将出现过的每一刻F,C的坐标和他们要走的偏移记录下来，如果未来又出现一样的，那么就是出现重复了，也就是后面就将陷入循环了；
//
//struct Bean{
//    int currentDirIndex = 0;
//    int currentDir[2] = {0,-1};
//    int currentPoint[2];
//};
//// 二维数组承接  （如何判定永远不会相遇,这个先不管，不好进行操作）  (dx,dy)表示下一次的偏移
//// 如果产生偏移时的旋转数组 每个元素为{x方向，y方向}
//int offset[4][2] = {{0,-1},
//                    {1,0},
//                    {0,1},
//                    {-1,0}};
//
//char map[10][10];
//int cnt;    // 次数
//
//// 检查是否可以移动到指定位置
//bool canMove(int x, int y) {
//    return x >= 0 && x < 10 && y >= 0 && y < 10 && map[x][y] != '*';
//}
//
//int main(){
//    // 初始化地图和角色
//    Bean F = Bean();
//    Bean C = Bean();
//    for(int i=0;i<10;i++){
//        for (int j = 0; j < 10; ++j) {
//            cin >> map[i][j];
//            if(map[i][j] == 'F'){
//                F.currentPoint[0] = i;
//                F.currentPoint[1] = j;
//            }else if(map[i][j] == 'C'){
//                C.currentPoint[0] = i;
//                C.currentPoint[1] = j;
//            }
//        }
//    }
//
//    // 记录历史状态，避免无限循环
//    bool visited[10][10][10][10][4][4] = {false};
//
//    while (1){
//        // 记录当前状态
//        int fx = F.currentPoint[0], fy = F.currentPoint[1];
//        int cx = C.currentPoint[0], cy = C.currentPoint[1];
//        int fd = F.currentDirIndex, cd = C.currentDirIndex;
//
//        // 检查是否出现循环
//        if (visited[fx][fy][cx][cy][fd][cd]) {
//            cout << 0 << endl;  // 永远不会相遇
//            return 0;
//        }
//        visited[fx][fy][cx][cy][fd][cd] = true;
//
//        // 计算下一个位置
//        int nextFx = fx + F.currentDir[0];
//        int nextFy = fy + F.currentDir[1];
//
//        if (!canMove(nextFx, nextFy)) {
//            // 不能前进，右转
//            F.currentDirIndex = (F.currentDirIndex+1)%4;
//            F.currentDir[0] = offset[F.currentDirIndex][0];
//            F.currentDir[1] = offset[F.currentDirIndex][1];
//        } else {
//            // 正常前行
//            F.currentPoint[0] = nextFx;
//            F.currentPoint[1] = nextFy;
//        }
//
//        // 计算牛的下一个位置
//        int nextCx = cx + C.currentDir[0];
//        int nextCy = cy + C.currentDir[1];
//
//        if (!canMove(nextCx, nextCy)) {
//            // 不能前进，右转
//            C.currentDirIndex = (C.currentDirIndex+1)%4;
//            C.currentDir[0] = offset[C.currentDirIndex][0];
//            C.currentDir[1] = offset[C.currentDirIndex][1];
//        } else {
//            // 正常前行
//            C.currentPoint[0] = nextCx;
//            C.currentPoint[1] = nextCy;
//        }
//
//        cnt++;
//
//        // 检查是否相遇
//        if(F.currentPoint[0] == C.currentPoint[0] && F.currentPoint[1] == C.currentPoint[1]){
//            break;
//        }
//    }
//    cout << cnt;
//
//
//    return 0;
//}

#include <iostream>
using namespace std;

struct Bean{
    int currentDirIndex = 0;
    int currentDir[2] = {0,-1};
    int currentPoint[2];
};

int offset[4][2] = {{0,-1},
                    {1,0},
                    {0,1},
                    {-1,0}};

char map[10][10];
int cnt = 0;  // 初始化计数器

// 检查是否可以移动到指定位置
bool canMove(int x, int y) {
    return x >= 0 && x < 10 && y >= 0 && y < 10 && map[x][y] != '*';
}

int main(){
    Bean F = Bean();
    Bean C = Bean();

    for(int i=0; i<10; i++){
        for (int j = 0; j < 10; ++j) {
            cin >> map[i][j];
            if(map[i][j] == 'F'){
                F.currentPoint[0] = i;
                F.currentPoint[1] = j;
            }else if(map[i][j] == 'C'){
                C.currentPoint[0] = i;
                C.currentPoint[1] = j;
            }
        }
    }

    // 记录历史状态，避免无限循环
    bool visited[10][10][10][10][4][4] = {false};

    while (true){
        // 检查当前状态是否已访问过
        int fx = F.currentPoint[0], fy = F.currentPoint[1];
        int cx = C.currentPoint[0], cy = C.currentPoint[1];
        int fd = F.currentDirIndex, cd = C.currentDirIndex;

        if (visited[fx][fy][cx][cy][fd][cd]) {
            cout << 0 << endl;  // 永远不会相遇
            return 0;
        }

        // 标记当前状态为已访问
        visited[fx][fy][cx][cy][fd][cd] = true;

        // 计算F的下一个位置
        int nextFx = fx + F.currentDir[0];
        int nextFy = fy + F.currentDir[1];

        if (!canMove(nextFx, nextFy)) {
            // 不能前进，右转
            F.currentDirIndex = (F.currentDirIndex+1)%4;
            F.currentDir[0] = offset[F.currentDirIndex][0];
            F.currentDir[1] = offset[F.currentDirIndex][1];
        } else {
            // 正常前行
            F.currentPoint[0] = nextFx;
            F.currentPoint[1] = nextFy;
        }

        // 计算牛的下一个位置
        int nextCx = cx + C.currentDir[0];
        int nextCy = cy + C.currentDir[1];

        if (!canMove(nextCx, nextCy)) {
            // 不能前进，右转
            C.currentDirIndex = (C.currentDirIndex+1)%4;
            C.currentDir[0] = offset[C.currentDirIndex][0];
            C.currentDir[1] = offset[C.currentDirIndex][1];
        } else {
            // 正常前行
            C.currentPoint[0] = nextCx;
            C.currentPoint[1] = nextCy;
        }

        cnt++;

        // 检查是否相遇
        if(F.currentPoint[0] == C.currentPoint[0] && F.currentPoint[1] == C.currentPoint[1]){
            cout << cnt << endl;
            return 0;
        }
    }

    return 0;
}

File: LuoGu/Part9/P1518.py
# 农夫和牛的实体类
class Bean:
    def __init__(self):
        self.current_dir_index = 0
        self.current_point = [0, 0]

# 移动的偏移量
offset = [[-1, 0],  # 北（上）
          [0, 1],   # 东（右）
          [1, 0],   # 南（下）
          [0, -1]]  # 西（左）

# 地图
game_map = [['' for _ in range(10)] for _ in range(10)]

# 记录历史状态，避免无限循环
visited = {}
cnt = 0

# 检查是否可以移动到指定位置的函数
def can_move(x, y):
    return 0 <= x < 10 and 0 <= y < 10 and game_map[x][y] != '*'

def main():
    global cnt
    F = Bean()
    C = Bean()

    # 读取地图
    for i in range(10):
        line = input().strip()
        for j in range(10):
            game_map[i][j] = line[j]
            if game_map[i][j] == 'F':
                F.current_point[0] = i
                F.current_point[1] = j
            elif game_map[i][j] == 'C':
                C.current_point[0] = i
                C.current_point[1] = j

    # 检查初始状态是否已经相遇
    if F.current_point[0] == C.current_point[0] and F.current_point[1] == C.current_point[1]:
        print(0)
        return

    while True:
        # 移动F
        next_fx = F.current_point[0] + offset[F.current_dir_index][0]
        next_fy = F.current_point[1] + offset[F.current_dir_index][1]

        if not can_move(next_fx, next_fy):
            # 不能前进，顺时针转90度
            F.current_dir_index = (F.current_dir_index + 1) % 4
        else:
            # 正常前行
            F.current_point[0] = next_fx
            F.current_point[1] = next_fy

        # 移动牛
        next_cx = C.current_point[0] + offset[C.current_dir_index][0]
        next_cy = C.current_point[1] + offset[C.current_dir_index][1]

        if not can_move(next_cx, next_cy):
            # 不能前进，顺时针转90度
            C.current_dir_index = (C.current_dir_index + 1) % 4
        else:
            # 正常前行
            C.current_point[0] = next_cx
            C.current_point[1] = next_cy

        cnt += 1

        # 检查是否相遇
        if F.current_point[0] == C.current_point[0] and F.current_point[1] == C.current_point[1]:
            print(cnt)
            return

        # 检查当前状态是否已访问过（在移动之后检查）
        fx, fy = F.current_point[0], F.current_point[1]
        cx, cy = C.current_point[0], C.current_point[1]
        fd, cd = F.current_dir_index, C.current_dir_index
        
        state = (fx, fy, cx, cy, fd, cd)
        
        if state in visited:
            print(0)  # 永远不会相遇
            return
        
        # 标记当前状态为已访问
        visited[state] = True

if __name__ == "__main__":
    main() 